####################################################################################
#                       FGa1: RNA-SEQ & GIBBS SAMPLING
####################################################################################

#setwd("~/Desktop/code/CompBio MPhil/Functional Genomics/Assignment1")

library(tidyverse)
library(cowplot)
set.seed(23456) #set random seed for replicating results


####################################################################################
#                             simulate.data
#implement an R functino called simulate.data that, given alpha1, ß1, apha 2, ß2, J and K 
#generates a randome vector of read counts from the model above. 
#the elements of the vetor should be oredered by experiment orered within cell line ordered within lab
#a = shape, b = rate
############################################################################

simulate.data <- function(beta.1, beta.2,alpha.1, alpha.2, J, K, I){
  #initialise vectors and matrices
  final.vector <- rep(0, I*J*K)
  final.array <- matrix(0, nrow=K, ncol=I*J)
  lambdas <- matrix(0, nrow=J, ncol=I)
  reads.1 <- matrix(0, nrow=K, ncol=J)
  reads.2 <- matrix(0, nrow=K, ncol=J)
  lambdas[1:J, 1] <- rgamma(J, shape=alpha.1, rate=beta.1) #lambda 1 on col 1
  lambdas[1:J, 2] <- rgamma(J, shape=alpha.2, rate=beta.2) #lambda 2 on col 2
  for (j in 1:J){
    reads.1[1:K,j] <- rpois(K, lambdas[j,1]) 
    reads.2[1:K,j] <- rpois(K, lambdas[j,2])
  }
  #final.vector <- c(reads.1, reads.2)
  reads <- list(reads.1, reads.2)
  return(reads)
}

trial1.vectorised <- simulate.data(beta.1 = 2, beta.2= 3, alpha.1 = 100, alpha.2=100, J=10, K=10, I=2)


#plot it against the indices of the
#elements in the vector, using a different point shape for each laboratory and a
#different point colour for each of the 2 × J experiments

plot.sim.data <- function(simulated.data){
  ### a function to plot simulated data using differnet point shapes for each lab
  ### and different point colour for each of 2xJ experiments
  # arg: simulated.data = a one dimentional vector with all data oredered by experiment
  #orered within cell line ordered within lab
  index <- 1:length(simulated.data)
  simulated.data.df <- data.frame("values" = simulated.data)
  simulated.data.df$colors <- rep(1:10, each = 10, times=2)
  simulated.data.df$lab <- c(rep(1,length(simulated.data)/2), rep(2,length(simulated.data)/2))
  
  #remember to change the values of parameters if plotting with different simulate data values
  plt1 <- ggplot(simulated.data.df, aes(x = index, y = values, colour = as.factor(colors),shape=as.factor(lab))) +
    geom_point(show.legend = T, size=3 ) + 
    ggtitle(expression(paste("Simulated data: ", beta[1]," = 2, ", beta[2], " = 3, ", alpha[1], " = ", 
                             alpha[2], " = 100, ", " J = 10, ", "K = 10, ", " I = 2" ))) +
    labs(x="Index",y="Number of reads", colour="Matched cell line", shape="Lab") + theme_bw() + 
    theme(plot.title = element_text ( lineheight=2, face="bold", color="black", size=20, hjust = 0.5 ),
          legend.position = "bottom", 
          legend.text = element_text(size = rel(1.1)),
          axis.title.y = element_text(size = rel(1.1)),
          axis.title.x = element_text(size = rel(1.1)),
          title = element_text(size=rel(1.5)),
          axis.text = element_text(size = rel(1.1)))
  
  plt1
}

simulated.data.1 <- unlist(trial1.vectorised)
plot.sim.data(simulated.data.1)


####################################################################
#Plot a set of 100 Poisson mass functions in a 10 × 10 grid for which the rate
#parameters have been randomly generated according to the gamma distribution
#shown in Equation 2, whose parameters have, in turn, been generated by the
#distribution in Equation 3. Use the same limits for all the x-axes.
########################################################################


pois.dist <- rep(list(rep(0,100)),100) #initialise the list of poisson distributions
for (ii in 1:100){
  beta <- rgamma(1, shape=0.5, rate=0.1)
  lambda <- rgamma(1, shape=100, rate=beta)
  pois.dist[[ii]] <- dpois(0:300,lambda)
}

par(mfrow=c(10,10)) #set 10 by 10 plots
par(mar=c(1,1,1,1)) # set margins
for (plt in 1:length(pois.dist)){
  plot(pois.dist[[plt]], type="l", ylim=c(0,0.12))
}



##################################################################
#                           GIBBS SAMPLER
################################################################
gibbs.sampler <- function(beta.1.init, beta.2.init,itn, y, I=2, J=10, K=10,alpha=100, a=0.5, b=0.1 ){
  ###
  #gibbs.sampler: take
  #beta.1.init: initial beta.1 at beginning of sampling 
  #beta.2 init: initial beta.2 at beginning of sampling
  #int: number of iterations
  #y: list of read counts created by simulate data. 
  #   1st element is reads from lab 1, 2nd element is reads from lab 2
  ###
  #select reads 1 and 2 from simulated data 
  reads.1 <- y[[1]] 
  reads.2 <- y[[2]]
  #initialise matrices for 
  beta.matrix <- matrix(0, nrow=I,ncol=itn) 
  lambda.matrix <- matrix(0, nrow=I*J, ncol=itn)
  beta.matrix[,1] <- c(beta.1.init,beta.2.init) #initialised matrix
  lambda.matrix[,1] <- rgamma(2*J, shape=100, rate=0.1)
  for (t in 2:itn){
    print(paste("Iteration", t))
    for (j in 1:J){
      lambda.matrix[j,t] <- rgamma(1, shape=(alpha+sum(reads.1[,j])), rate=(beta.matrix[1,(t-1)] + K))
      lambda.matrix[(J+j), t] <-rgamma(1, shape=(alpha+sum(reads.2[,j])), rate=(beta.matrix[2,(t-1)] + K))
    }
    beta.matrix[1,t] <- rgamma(1, shape=(a+alpha*J), rate=(b+ sum(lambda.matrix[(1:J),t])))
    beta.matrix[2,t] <- rgamma(1, shape=(a+alpha*J), rate=(b+sum(lambda.matrix[(J+1):(2*J),t])))   
  }
  return(list(lambda.matrix, beta.matrix))
}

#create the read counts from simulate data with beta.1 as 0.12, beta.2 as 0.1
readcounts <- simulate.data(beta.1 = 0.12, beta.2= 0.10, alpha.1 = 100, alpha.2=100, J=10, K=10, I=2)

#run the gibbs sampler with 10000 iterations, initial betas1 and 2 as =4 and =5 and readcounts from simulate data as y
tenthousand <- gibbs.sampler(beta.1.init=4, beta.2.init=5,itn=10000, y=readcounts)

#select sampled betas from output of gibbs sampler
beta.1 <- tenthousand[[2]][1, ]; mean(beta.1)
beta.2 <- tenthousand[[2]][2, ]; mean(beta.2)

#create dataframe for plotting 
beta.df <- data.frame("beta1" = as.numeric(beta.1) , "beta2" = as.numeric(beta.2))

#reset plotting parameters
par(mfrow=c(1,1)) #set back to 1 plot
par(mar=c(10,10,10,10)) # set margins

betas.plot <- ggplot(beta.df, aes(x=(1:dim(beta.df)[1]))) + geom_line(aes(y=beta.df$beta1, colour="red")) +
  geom_line(aes(y=beta.df$beta2, colour="green")) + coord_cartesian(ylim = (0.11)) +
  geom_hline(yintercept=0.12, color = "black") + geom_hline(yintercept=0.10, color = "black") + 
  ggtitle(expression(paste("Real vs sampled ", beta[1], " and ", beta[2]))) + 
  xlab("Iteration number") + ylab("Value of betas as by Gibbs sampler") +
  scale_colour_discrete(guide = guide_legend(reverse=TRUE),name="Betas", breaks=c("blue","pink"), 
                        labels=c(expression(paste(beta[2], " = 0.10")), expression(paste(beta[1], " = 0.12")))) +
  theme_bw() + theme(plot.title = element_text ( lineheight=2, face="bold", color="black", size=20, hjust = 0.5 ),
                     legend.position = "right", 
                     legend.text = element_text(size = rel(1.1)),
                     axis.title.y = element_text(size = rel(1.1)),
                     axis.title.x = element_text(size = rel(1.1)),
                     title = element_text(size=rel(1.5)),
                     axis.text = element_text(size = rel(1.1)))

betas.plot


########################################################################  
#plot a histogram of Gibbs trace of alpha1/beta1 - alpha2/beta2 
#after discarding the balues for first 100 iterations
########################################################################

#discarding first 100 iterations
clean.beta.1 <- beta.1[101:10000]
clean.beta.2 <- beta.2[101:10000]

#alpha1 and alpha2 both set as 100 
mean.gibbs.diff <- c((100/clean.beta.1)-(100/clean.beta.2))

par(mar=c(4,4,4,4))
par(mfrow=c(1,1))
h <- hist(mean.gibbs.diff, breaks=30, main=expression(paste("Histogram of Gibbs trace of ", 
                                                       frac(alpha[1],beta[1]), " - ", 
                                                       frac(alpha[2],beta[2]))),
          xlim=c(-400,50), xlab=expression(paste(frac(alpha[1],beta[1]), " - ", 
                                                 frac(alpha[2],beta[2]))))


## encoding!! can save as specific encoding 

#############################################################################################
#                   probability that lab 2 generates more reads than lab 1:
###############################################################################################

shapiro.test(sample((100/clean.beta.1), 5000)) 
shapiro.test(sample((100/clean.beta.2), 5000)) 

mean.posteriors <- data.frame("mean.posterior.1" = (100/clean.beta.1), "mean.posterior.2" = (100/clean.beta.2) )
wilcox.test(mean.posterior.1, mean.posterior.2, data=mean.posteriors)


#check if distribution of difference/residuals is normal
shapiro.test(sample(mean.gibbs.diff, 5000))
#p value: 0.363 --> distribution is normal 
mean(mean.gibbs.diff); sd(mean.gibbs.diff)
#compare residuals distribution to made up normal distribution mean 0 sd same as residuals
std.meangibbs.diff <- sd(mean.gibbs.diff)
x<- rnorm(9900, mean=0, sd=std.meangibbs.diff)
t.test(x, mean.gibbs.diff)


###############################################################################################################
#                                   Comparing effects of J and K
#for each pair (J,K) {10,100}x{10,100} plot a pair of histograms depicting posteior distributions of ß1 and ß2.
#comment on the effects of J and K on posteiror variances of these parameters
###############################################################################################################
J10K10.reads <- simulate.data(beta.1 = 0.12, beta.2= 0.1, alpha.1 = 100, alpha.2=100, J=10, K=10, I=2)
J100K10.reads <-  simulate.data(beta.1 = 0.12, beta.2= 0.1, alpha.1 = 100, alpha.2=100, J=100, K=10, I=2)
J10K100.reads <- simulate.data(beta.1 = 0.12, beta.2= 0.1, alpha.1 = 100, alpha.2=100, J=10, K=100, I=2)
J100K100.reads <- simulate.data(beta.1 = 0.12, beta.2= 0.1, alpha.1 = 100, alpha.2=100, J=100, K=100, I=2)


J10K10.gibbs <- gibbs.sampler(beta.1.init=0.12, beta.2.init=0.1,itn=10000, y=J10K10.reads, J=10, K=10)
J100K10.gibbs <- gibbs.sampler(beta.1.init=0.12, beta.2.init=0.1,itn=10000, y=J100K10.reads, J=100, K=10)
J10K100.gibbs <- gibbs.sampler(beta.1.init=0.12, beta.2.init=0.1,itn=10000, y=J10K100.reads, J=10, K=100)
J100K100.gibbs <- gibbs.sampler(beta.1.init=0.12, beta.2.init=0.1,itn=10000, y=J100K100.reads, J=100, K=100)

J10K10.beta.1 <- J10K10.gibbs[[2]][1,]; var(J10K10.beta.1)
J10K10.beta.2 <- J10K10.gibbs[[2]][2,]; var(J10K10.beta.2)
J10K10 <- c(J10K10.beta.1,J10K10.beta.2)
betas1010 <- c(rep(1, length(J10K10.beta.1)), rep(2, length(J10K10.beta.2)))
J10K10df <- data.frame('values' = J10K10,'betas'=betas1010)

J100K10.beta.1 <- J100K10.gibbs[[2]][1,]; var(J100K10.beta.1)
J100K10.beta.2 <- J100K10.gibbs[[2]][2,]; var(J100K10.beta.2)
J100K10 <- c(J100K10.beta.1,J100K10.beta.2)
betas10010 <- c(rep(1, length(J100K10.beta.1)), rep(2, length(J100K10.beta.2)))
J100K10df <- data.frame('values' = J100K10,'betas'=betas10010)


J10K100.beta.1 <- J10K100.gibbs[[2]][1,]; var(J10K100.beta.1)
J10K100.beta.2 <- J10K100.gibbs[[2]][2,]; var(J10K100.beta.2)
J10K100 <- c(J10K100.beta.1,J10K100.beta.2)
betas10100 <- c(rep(1, length(J10K100.beta.1)), rep(2, length(J10K100.beta.2)))
J10K100df <- data.frame('values' = J10K100,'betas'=betas10100)


J100K100.beta.1 <- J100K100.gibbs[[2]][1,]; var(J100K100.beta.1)
J100K100.beta.2 <- J100K100.gibbs[[2]][2,]; var(J100K100.beta.2)
J100K100 <- c(J100K100.beta.1,J100K100.beta.2)
betas100100 <- c(rep(1, length(J100K100.beta.1)), rep(2, length(J100K100.beta.2)))
J100K100df <- data.frame('values' = J100K100,'betas'=betas100100)





p1 <- ggplot(J10K10df, aes(x=values, color=as.factor(betas1010))) + geom_histogram(bins=100) +
  coord_cartesian(xlim = (0.11))+ ylim(0,1100) + ggtitle("J = 10, K = 10") + labs(y="Freqency", color="Betas") + theme_bw() + 
  scale_colour_discrete(name="Betas", breaks=c("1","2"), labels=c(expression(beta[1]), expression(beta[2]))) +
  theme(plot.title = element_text ( lineheight=2, face="bold", color="black", size=20, hjust = 0.5 ),
        legend.position = "right", 
        legend.text = element_text(size = rel(1.1)),
        axis.title.y = element_text(size = rel(1.1)),
        axis.title.x = element_blank(),
        title = element_text(size=rel(1.5)),
        axis.text = element_text(size = rel(1.1)))
  
  
p2 <- ggplot(J100K10df,aes(x=values, color=as.factor(betas10010) )) + geom_histogram(bins=100) +
  coord_cartesian(xlim = (0.11))+ ylim(0,1100) + ggtitle("J = 100, K = 10") + labs(y="Freqency", color="Betas") + theme_bw() + 
  scale_colour_discrete(name="Betas", breaks=c("1","2"), labels=c(expression(beta[1]), expression(beta[2]))) +
  theme(plot.title = element_text ( lineheight=2, face="bold", color="black", size=20, hjust = 0.5 ),
        legend.position = "right", 
        legend.text = element_text(size = rel(1.1)),
        axis.title.y = element_text(size = rel(1.1)),
        axis.title.x = element_blank(),
        title = element_text(size=rel(1.5)),
        axis.text = element_text(size = rel(1.1)))
  
  
p3 <- ggplot(J10K100df,aes(x=values, color=as.factor(betas10100) )) + geom_histogram(bins=100) +
  coord_cartesian(xlim = (0.11)) + ylim(0,1100) + ggtitle("J = 10, K = 100") + labs(y="Freqency", color="Betas") + theme_bw() + 
  scale_colour_discrete(name="Betas", breaks=c("1","2"), labels=c(expression(beta[1]), expression(beta[2]))) +
  theme(plot.title = element_text ( lineheight=2, face="bold", color="black", size=20, hjust = 0.5 ),
        legend.position = "right", 
        legend.text = element_text(size = rel(1.1)),
        axis.title.y = element_text(size = rel(1.1)),
        axis.title.x = element_blank(),
        title = element_text(size=rel(1.5)),
        axis.text = element_text(size = rel(1.1)))

p4 <- ggplot(J100K100df,aes(x=values, color=as.factor(betas100100) )) + geom_histogram(bins=100) +
  coord_cartesian(xlim = (0.11)) + ylim(0,1100) + ggtitle("J = 100, K = 100") + labs(y="Freqency", color="Betas") + theme_bw() + 
  scale_colour_discrete(name="Betas", breaks=c("1","2"), labels=c(expression(beta[1]), expression(beta[2]))) +
  theme(plot.title = element_text ( lineheight=2, face="bold", color="black", size=20, hjust = 0.5 ),
        legend.position = "right", 
        legend.text = element_text(size = rel(1.1)),
        axis.title.y = element_text(size = rel(1.1)),
        axis.title.x = element_blank(),
        title = element_text(size=rel(1.5)),
        axis.text = element_text(size = rel(1.1)))


plot_grid(p1, p3,p2, p4, labels = "AUTO")


########################################################################################
#                                 Plot variability of Gibbs Sampler
########################################################################################

set.seed(12)

repeats <- 100
readcounts <- simulate.data(beta.1 = 0.12, beta.2= 0.10, alpha.1 = 100, alpha.2=100, J=10, K=10, I=2)
gibbs.beta.1 <- rep(0,repeats)
gibbs.beta.2 <- rep(0,repeats)
for (r in 1:repeats){
  hundred <- gibbs.sampler(beta.1.init=4, beta.2.init=5,itn=300, y=readcounts)
  beta.1 <- hundred[[2]][1, ]; mean(beta.1)
  beta.2 <- hundred[[2]][2, ]; mean(beta.2)
  gibbs.beta.1[r] <- mean(beta.1[100:300])
  gibbs.beta.2[r] <- mean(beta.2[100:300])
  
}

par(mfrow=c(1,2))
par(mar=c(3,3,3,3))
hist(gibbs.beta.1)
abline(v=0.12)
hist(gibbs.beta.2)
abline(v=0.10)


#select sampled betas from output of gibbs sampler


